lib/plox.ex

```diff
@@ -375,82 +375,54 @@ defmodule Plox do
   attr :rest, :global, include: @svg_presentation_globals
 
   def circle(assigns) do
-    assigns
-    |> assign(dataset: validate_zero_or_one_dataset(Map.take(assigns, [:cx, :cy, :r, :fill])))
-    |> do_circle()
-  end
-
-  defp do_circle(%{dataset: :none} = assigns) do
-    ~H"""
-    <circle cx={@cx} cy={@cy} r={@r} fill={@fill} {@rest} />
-    """
-  end
-
-  defp do_circle(%{dataset: _dataset} = assigns) do
     ~H"""
     <circle
-      :for={data_point <- @dataset.data}
-      cx={maybe_graph(@cx, data_point)}
-      cy={maybe_graph(@cy, data_point)}
-      r={maybe_graph(@r, data_point)}
-      fill={maybe_graph(@fill, data_point)}
-      stroke={maybe_graph(@stroke, data_point)}
-      stroke-width={maybe_graph(assigns[:"stroke-width"], data_point)}
+      :for={{cx, cy, r, fill, stroke, stroke_width} <- values([@cx, @cy, @r, @fill, @stroke, assigns[:"stroke-width"]])}
+      cx={cx}
+      cy={cy}
+      r={r}
+      fill={fill}
+      stroke={stroke}
+      stroke-width={stroke_width}
       {@rest}
     />
     """
   end
 
-  defp validate_zero_or_one_dataset(assigns) do
-    datasets =
-      assigns
-      |> Enum.flat_map(fn
-        {_key, %Plox.DatasetAxis{} = dataset_axis} -> [dataset_axis.dataset]
-        {_key, _} -> []
-      end)
-      |> Enum.uniq()
-
-    case datasets do
+  defp validate_zero_or_one_dataset(data) do
+    data
+    |> Enum.flat_map(fn
+      %Plox.DatasetAxis{} = dataset_axis -> [dataset_axis.dataset]
+      _ -> []
+    end)
+    |> Enum.uniq()
+    |> case do
       [] -> :none
       [dataset] -> dataset
```

Comment:
```suggestion
      [_dataset] -> :ok
```
we no longer need a reference to the dataset here.